src/runtime/ndarray.cc

```diff
@@ -51,91 +48,61 @@ inline size_t GetDataSize(const DLTensor& arr) {
-  static NDArray Create(std::vector<int64_t> shape,
-                        DLDataType dtype,
-                        DLContext ctx) {
-    VerifyDataType(dtype);
-    // critical zone
-    NDArray::Container* data = new NDArray::Container();
-    data->deleter = DefaultDeleter;
-    NDArray ret(data);
-    ret.data_ = data;
-    // RAII now in effect
-    // setup shape
-    data->shape_ = std::move(shape);
-    data->dl_tensor.shape = dmlc::BeginPtr(data->shape_);
-    data->dl_tensor.ndim = static_cast<int>(data->shape_.size());
-    // setup stride (this should be optional, but some framework
-    //   does not support NULL stride and thus will crash the program).
-    data->stride_.resize(data->dl_tensor.ndim, 1);
-    for (int i = data->dl_tensor.ndim - 2; i >= 0; --i) {
-      data->stride_[i] = data->shape_[i+1] * data->stride_[i+1];
-    }
-    data->dl_tensor.strides = dmlc::BeginPtr(data->stride_);
-    // setup dtype
-    data->dl_tensor.dtype = dtype;
-    // setup ctx
-    data->dl_tensor.ctx = ctx;
-    return ret;
-  }
-  // Implementation of API function
-  static DLTensor* MoveAsDLTensor(NDArray arr) {
-    DLTensor* tensor = reinterpret_cast<DLTensor*>(arr.data_);
-    CHECK(tensor == const_cast<DLTensor*>(arr.operator->()));
-    arr.data_ = nullptr;
-    return tensor;
+    dgl::runtime::DeviceAPI::Get(ptr->dl_tensor.ctx)->FreeDataSpace(
+        ptr->dl_tensor.ctx, ptr->dl_tensor.data);
   }
-  // Container to DLManagedTensor
-  static DLManagedTensor* ToDLPack(NDArray::Container* from) {
-    CHECK(from != nullptr);
-    DLManagedTensor* ret = new DLManagedTensor();
-    ret->dl_tensor = from->dl_tensor;
-    ret->manager_ctx = from;
-    from->IncRef();
-    ret->deleter = NDArrayDLPackDeleter;
-    return ret;
+  delete ptr;
+}
+
+NDArray NDArray::Internal::Create(std::vector<int64_t> shape,
+                                  DGLDataType dtype, DGLContext ctx) {
+  VerifyDataType(dtype);
+  // critical zone
+  NDArray::Container* data = new NDArray::Container();
+  data->deleter = DefaultDeleter;
+  NDArray ret(data);
+  ret.data_ = data;
+  // RAII now in effect
+  // setup shape
+  data->shape_ = std::move(shape);
+  data->dl_tensor.shape = dmlc::BeginPtr(data->shape_);
+  data->dl_tensor.ndim = static_cast<int>(data->shape_.size());
+  // setup stride (this should be optional, but some framework
+  //   does not support NULL stride and thus will crash the program).
+  data->stride_.resize(data->dl_tensor.ndim, 1);
+  for (int i = data->dl_tensor.ndim - 2; i >= 0; --i) {
+    data->stride_[i] = data->shape_[i+1] * data->stride_[i+1];
   }
-};
+  data->dl_tensor.strides = dmlc::BeginPtr(data->stride_);
+  // setup dtype
+  data->dl_tensor.dtype = dtype;
+  // setup ctx
+  data->dl_tensor.ctx = ctx;
+  return ret;
+}
+
+DGLArray* NDArray::Internal::MoveAsDGLArray(NDArray arr) {
+  DGLArray* tensor = reinterpret_cast<DGLArray*>(arr.data_);
+  CHECK(tensor == const_cast<DGLArray*>(arr.operator->()));
```

Comment:
For L101-102, can we just do
```c++
DGLArray* tensor = arr.data_->dl_tensor?
```

Reply:
It's a bit tricky here. We use `reinterpret_cast` here because we want to keep this `NDArray::Container` object (but treat its former part as a DGLArray), so that when a `DGLArray` object is passed from Python to C++, we can `reinterpret_cast` it back to an `NDArray::Container`, for example, in https://github.com/dmlc/dgl/blob/36dbccee1213653b7e4d1a07b6941b0f62d203c5/src/runtime/ndarray.cc#L499-L502
This is also why we don't free `arr.data_`.

I don't know why it looks like this at the beginning, but this is how it behaves/works now...

Reply:
I see. The design right now is complicated. You change is fine as it doesn't modify the original behavior. Let's simplify it later in another PR.

Reply:
I think the core issue is still that `NDArray::Container` is coupled too tightly with `DLManagedTensor`. In the ideal case, `DLManagedTensor` should be transient object whose lifescycle is limited to the DLPack data exchange.